Implement attributes for UpdateFields

killerwife opened the pull request:
Prototype for improving parsing of update fields. Initial addition of enum attributes for the most used ones. Improves readability of the actual sniff. Gets rid of some duplicate code and enables future extensibility for stuff like unitflags or other enum based fields.

Warpten commented on WowPacketParser/Parsing/Parsers/UpdateHandler.cs:
```diff
@@ -172,175 +180,336 @@ public static void ReadObjectsBlock(Packet packet, object index)
+                }
+                UpdateFieldType updateFieldType = UpdateFieldType.Default;
+
+                List<UpdateField> fieldData = new List<UpdateField>();
```
Meh. I see why, and don't see a better approach

Warpten commented on WowPacketParser/Parsing/Parsers/UpdateHandler.cs:
```diff
@@ -172,175 +180,336 @@ public static void ReadObjectsBlock(Packet packet, object index)
+                        }
+                        break;
+                    }
+                    case UpdateFieldType.Custom:
```
Again, I don't like splitting implementation from definition here, especially for corner cases/hacks, but besides shoving a delegate in the attribute, idk

killerwife replied:
There will be plenty, for example parsing of unitflags, or parsing of bytes 0, bytes1 and so on, for which a delegate should be provided based on the version, but at the moment I did not want to spend time on it, since this on its own is a huge improvement already.

Shauren commented on WowPacketParser/Parsing/UpdateFieldAttributes.cs:
```diff
@@ -0,0 +1,36 @@
+        Default, // Old formatting - Uint/Float - Supports variable length
+        Guid, // Must be 64-bit or 128-bit
+        Quaternion, // 4x float
+        PackedQuaternion, // 2x float
```
Comment incorrect, packed quat is a single uint64

Shauren commented on WowPacketParser/Parsing/Parsers/UpdateHandler.cs:
```diff
@@ -172,175 +178,338 @@ public static void ReadObjectsBlock(Packet packet, object index)
-                updateMask = new int[maskSize];
-                for (var i = 0; i < maskSize; i++)
-                    updateMask[i] = packet.ReadInt32();
+                var memInfo = fieldType.GetMember(key);
```
reflection use must be somehow cached, otherwise performance will be horrible

killerwife replied:
How do you suggest any such caching?

Shauren replied:
I think altering the updatefield storage would work (not 100% sure yet)

Shauren replied:
Also I was including cache of GetCustomAttribute in there too

killerwife replied:
Well let me know, because I barely know reflection. The other stuff is valid and I can address, but not this.

Warpten replied:
Is that overall runtime or just update object? If the latter then don't bother increasing code complexity by adding caching, this opcode is too much of a shit show already imo

killerwife replied:
Overall.

Shauren commented on WowPacketParser/Parsing/Parsers/UpdateHandler.cs:
```diff
@@ -172,175 +178,338 @@ public static void ReadObjectsBlock(Packet packet, object index)
+                            if (mask[start + k] && (!isCreating || fieldData[k].UInt32Value != 0))
+                            {
+                                byte[] intBytes = BitConverter.GetBytes(fieldData[k].UInt32Value);
+                                if (BitConverter.IsLittleEndian)
```
dont reverse, byte order on network never changes

Shauren commented on WowPacketParser/Parsing/Parsers/UpdateHandler.cs:
```diff
@@ -172,175 +178,338 @@ public static void ReadObjectsBlock(Packet packet, object index)
-                        if (!fieldMask[j])
-                            continue;
+                var flag = packet.ReadUInt16();
+                var cnt = flag & 0x7FFF;
```
broken. depending on version its either uint16 flag + uint32 size OR uint8 flag + uint16 size (earlier versions)

Shauren commented on WowPacketParser/Parsing/Parsers/UpdateHandler.cs:
```diff
@@ -2554,7 +2723,9 @@ private static MovementInfo ReadMovementUpdateBlock(Packet packet, WowGuid guid,
                 var moveFlags = moveInfo.Flags;
 
-                for (var i = 0; i < 9; ++i)
+                var speeds = ClientVersion.AddedInVersion(ClientVersionBuild.V3_0_3_9183) ? 9 : 8;
```
seems like a strange thing to change between 3.0.2 and 3.0.3 tbh

killerwife replied:
Not sure if this was intended for 3.0.2 vs 3.0.3 as more TBC vs WOTLK.

Shauren commented on WowPacketParser/Parsing/Parsers/UpdateHandler.cs:
```diff
@@ -2604,6 +2775,14 @@ private static MovementInfo ReadMovementUpdateBlock(Packet packet, WowGuid guid,
                             packet.ReadSingle("Final Spline Orientation", index);
                         else if (splineFlags.HasAnyFlag(SplineFlag.FinalPoint))
                             packet.ReadVector3("Final Spline Coords", index);
+                        else if (splineFlags.HasAnyFlag(SplineFlag.CatmullRom))
```
this is definitely not there for 3.3.5 (no idea about other wotlk builds)

Shauren commented on WowPacketParser/Parsing/Parsers/UpdateHandler.cs:
```diff
@@ -2604,6 +2775,14 @@ private static MovementInfo ReadMovementUpdateBlock(Packet packet, WowGuid guid,
+
+                        if (ClientVersion.RemovedInVersion(ClientVersionBuild.V3_1_1_9806))
+                        {
+                            if (splineFlags.HasAnyFlag(SplineFlag.FinalOrientation))
```
definitely not correct (i dont even know the structure) - you will attempt to read this twice for builds < 3.1.1

killerwife replied:
Will doublecheck this one. I am taking this info from our old parser, which at least works for these versions.

Shauren commented on WowPacketParser/Enums/UpdateFields.cs:
```diff
@@ -142,30 +148,46 @@ public enum UnitField
         UNIT_FIELD_EFFECTIVE_LEVEL,
         UNIT_FIELD_END,
+        [UpdateField(UpdateFieldType.Uint)]
+        [UpdateField(UpdateFieldType.Custom, ClientVersionBuild.V6_2_3_20726)]
```
not sure why you would flag this as custom only in 6.2.3+, its always been a reference to FactionTemplate.dbc/db2

killerwife replied:
Because this hack was only in WOD+ update handler. So I preserved old behaviour.

Warpten commented on WowPacketParser/Parsing/Parsers/UpdateHandler.cs:
```diff
@@ -172,175 +178,338 @@ public static void ReadObjectsBlock(Packet packet, object index)
-                updateMask = new int[maskSize];
-                for (var i = 0; i < maskSize; i++)
-                    updateMask[i] = packet.ReadInt32();
+                var memInfo = fieldType.GetMember(key);
```
It's actually not too bad for enums. You can just construct an array of MemberInfo for the given type and reuse it. I think this is fine though.

killerwife replied:
It runs slightly slower than the master version. Talking maybe 10-20%. I have been using this ever since I made this PR.

Warpten commented on WowPacketParser/Parsing/Parsers/UpdateHandler.cs:
```diff
@@ -172,175 +178,338 @@ public static void ReadObjectsBlock(Packet packet, object index)
-                updateMask = new int[maskSize];
-                for (var i = 0; i < maskSize; i++)
-                    updateMask[i] = packet.ReadInt32();
+                var memInfo = fieldType.GetMember(key);
```
It's actually not too bad for enums. You can just construct an array of MemberInfo for the given type and reuse it. I think this is fine though.

killerwife commented:
@Shauren I did the following. I removed the catmullrom and the finalorientation thing, because atm I do not have time to delve into the client reversal. If you don't have a 3.0.2 to 3.0.8 sniff, let me know.

When it comes to reflection, the slowdown is not visible. You can test it yourself on any sniff version you'd like. It seems C# is doing optimizations already.

What do we do about the faction? Do you want me to set it to DBC reference for all versions? Does that mean you have to have some additional setup to have the field extracted? I only kept the original behaviour.

The flag stuff is fixed, I adjusted speeds count to 3.0.2.

killerwife commented:
Ok, I dealt with the DBC as well. Custom handling is a bit of an open book. The possible ideas are adding it to the enum, like lets say faction, instead of custom, or making an array of delegates based on enum, which does appropriate things. Should you want me to do something with the reflection, shoot some links my way. If you want me to post some performance numbers here, let me know. If not I can supply sniffs with which to check stuff.
